### ISSUE TITLE
 Query Result Set Very Slow Optimization Problem，optimizer Bug？

### ISSUE THREAD

**Issue by 444728248** [greenplum-db/gpdb#7118]
first ： I put the filters together, which is equivalent to my time going through the partition table, GID going through the index, and info_imsi_temp doing inner join.

Second ： I put the results of the screening criteria into a temporary table, and the last two temporary tables are inner join.

There is a big difference in performance: first, it takes more than 20 seconds, and second, it takes only two seconds.

**first** 
```
set optimizer = off;
EXPLAIN ANALYZE 
select a.* from info_imsi  a
							 inner join
							info_imsi_temp b
on a.gid = b.gid and a.imsi<>b.imsi and a.capturetime>=(b.capturetime-60) and a.capturetime<=(b.capturetime+60) 
where  a.capturetime>=1514767448 and a.capturetime<=1516767448


Gather Motion 8:1  (slice2; segments: 8)  (cost=145.08..430.71 rows=9 width=214)
  Rows out:  13046 rows at destination with 2878 ms to first row, 22429 ms to end.
  ->  Hash Join  (cost=145.08..430.71 rows=2 width=214)
        Hash Cond: a.gid = b.gid
        Join Filter: a.imsi::text <> b.imsi::text AND a.capturetime >= (b.capturetime - 60::numeric) AND a.capturetime <= (b.capturetime + 60::numeric)
        Rows out:  Avg 1630.8 rows x 8 workers.  Max 2044 rows (seg6) with 3048 ms to first row, 22157 ms to end.
        Executor memory:  2K bytes avg, 2K bytes max (seg0).
        Work_mem used:  2K bytes avg, 2K bytes max (seg0). Workfile: (0 spilling)
        (seg6)   Hash chain length 24.0 avg, 24 max, using 1 of 131072 buckets.
        ->  Append  (cost=134.28..418.39 rows=34 width=214)
              Rows out:  Avg 13542132.9 rows x 8 workers.  Max 13744503 rows (seg1) with 2550 ms to first row, 9588 ms to end.
              ->  Bitmap Append-Only Column-Oriented Scan on info_imsi_1_prt_p1 a  (cost=134.28..209.19 rows=17 width=214)
                    Recheck Cond: capturetime >= 1514767448::numeric AND capturetime <= 1516767448::numeric
                    Rows out:  Avg 13542132.9 rows x 8 workers.  Max 13744503 rows (seg1) with 2550 ms to first row, 8241 ms to end.
                    ->  Bitmap Index Scan on gid_capturetime_1_prt_p1  (cost=0.00..134.25 rows=17 width=0)
                          Index Cond: capturetime >= 1514767448::numeric AND capturetime <= 1516767448::numeric
                          Bitmaps out:  Avg 1.0 x 8 workers.  Max 1 (seg0) with 3124 ms to end, start offset by -1266 ms.
                          Work_mem used:  3398K bytes avg, 3449K bytes max (seg1).
              ->  Bitmap Append-Only Column-Oriented Scan on info_imsi_1_prt_other a  (cost=134.28..209.19 rows=17 width=214)
                    Recheck Cond: capturetime >= 1514767448::numeric AND capturetime <= 1516767448::numeric
                    Rows out:  0 rows (seg0) with 0.530 ms to end.
                    ->  Bitmap Index Scan on gid_capturetime_1_prt_other  (cost=0.00..134.25 rows=17 width=0)
                          Index Cond: capturetime >= 1514767448::numeric AND capturetime <= 1516767448::numeric
                          Bitmaps out:  Avg 1.0 x 8 workers.  Max 1 (seg0) with 0.017 ms to end, start offset by 19940 ms.
        ->  Hash  (cost=8.40..8.40 rows=24 width=31)
              Rows in:  Avg 24.0 rows x 8 workers.  Max 24 rows (seg0) with 0.092 ms to end, start offset by -1267 ms.
              ->  Broadcast Motion 8:8  (slice1; segments: 8)  (cost=0.00..8.40 rows=24 width=31)
                    Rows out:  Avg 24.0 rows x 8 workers at destination.  Max 24 rows (seg0) with 0.046 ms to first row, 0.069 ms to end.
                    ->  Append-only Columnar Scan on info_imsi_temp b  (cost=0.00..6.24 rows=3 width=31)
                          Rows out:  Avg 4.0 rows x 6 workers.  Max 8 rows (seg0) with 0.669 ms to first row, 0.727 ms to end.
Slice statistics:
  (slice0)    Executor memory: 554K bytes.
  (slice1)    Executor memory: 829K bytes avg x 8 workers, 829K bytes max (seg0).
  (slice2)    Executor memory: 8864K bytes avg x 8 workers, 8864K bytes max (seg0).  Work_mem: 3449K bytes max.
Statement statistics:
  Memory used: 128000K bytes
Settings:  effective_cache_size=20GB; optimizer=off; random_page_cost=0
Optimizer status: legacy query optimizer
Total runtime: 22432.522 ms
```
**Second** 
```
explain ANALYZE
CREATE TABLE info_imsi_temp1 as
SELECT * FROM info_imsi a where a.gid = 76 and a.capturetime>=1514767448 and a.capturetime<=1515445848

EXPLAIN ANALYZE 
select a.* from info_imsi_temp1 a 
 inner join
							info_imsi_temp b
on a.gid = b.gid and a.imsi<>b.imsi and a.capturetime>=(b.capturetime-60) and a.capturetime<=(b.capturetime+60) 

Gather Motion 8:1  (slice2; segments: 8)  (cost=10.80..4892080.68 rows=4465485 width=45)
  Rows out:  13046 rows at destination with 6.030 ms to first row, 2848 ms to end.
  ->  Hash Join  (cost=10.80..4892080.68 rows=558186 width=45)
        Hash Cond: a.gid = b.gid
        Join Filter: a.imsi::text <> b.imsi::text AND a.capturetime >= (b.capturetime - 60::numeric) AND a.capturetime <= (b.capturetime + 60::numeric)
        Rows out:  Avg 1630.8 rows x 8 workers.  Max 2044 rows (seg6) with 4.835 ms to first row, 2806 ms to end.
        Executor memory:  2K bytes avg, 2K bytes max (seg0).
        Work_mem used:  2K bytes avg, 2K bytes max (seg0). Workfile: (0 spilling)
        (seg6)   Hash chain length 24.0 avg, 24 max, using 1 of 262144 buckets.
        ->  Seq Scan on info_imsi_temp1 a  (cost=0.00..20394.95 rows=209325 width=45)
              Rows out:  Avg 210399.0 rows x 8 workers.  Max 214520 rows (seg5) with 0.102 ms to first row, 22 ms to end.
        ->  Hash  (cost=8.40..8.40 rows=24 width=31)
              Rows in:  Avg 24.0 rows x 8 workers.  Max 24 rows (seg0) with 0.078 ms to end, start offset by -1239 ms.
              ->  Broadcast Motion 8:8  (slice1; segments: 8)  (cost=0.00..8.40 rows=24 width=31)
                    Rows out:  Avg 24.0 rows x 8 workers at destination.  Max 24 rows (seg0) with 0.045 ms to first row, 0.059 ms to end.
                    ->  Append-only Columnar Scan on info_imsi_temp b  (cost=0.00..6.24 rows=3 width=31)
                          Rows out:  Avg 4.0 rows x 6 workers.  Max 8 rows (seg0) with 1.179 ms to first row, 1.232 ms to end.
Slice statistics:
  (slice0)    Executor memory: 386K bytes.
  (slice1)    Executor memory: 829K bytes avg x 8 workers, 829K bytes max (seg0).
  (slice2)    Executor memory: 4431K bytes avg x 8 workers, 4431K bytes max (seg0).  Work_mem: 2K bytes max.
Statement statistics:
  Memory used: 128000K bytes
Settings:  effective_cache_size=20GB; optimizer=off; random_page_cost=0
Optimizer status: legacy query optimizer
Total runtime: 2879.395 ms
```

**Comment by paul-guo-** [comment]
The plan format is a bit hard to read. Could you please use markdown format (typically use ```) to make it easier to read?

**Comment by hlinnaka** [comment]
How long did the CREATE TABLE AS command take? For a fair comparison, you'll have to take that into account, too.

In the second case, you added a `a.gid = 76` condition to the CREATE TABLE AS, which was not present in the first case.. You still got the same number of rows out, so I guess you knew beforehand that the only matching rows in 'b' have b.gid=76. The planner does not know that, so it cannot take advantage of that. It could perhaps have done the join the other way round: scan b first, and then fetch rows with matching gid from a. But I don't think it would've been faster.

In short, I don't think there's much we can do in the planner to improve this. Adding the `a.gid = 76` to the first query, too, might help. Closing.
